Add DeprecatedBlockTag rule

utybo opened the pull request:
<!-- A similar PR may already be submitted!
Please search among the [Pull requests](https://github.com/detekt/detekt/pulls) before creating one.

Thanks for submitting a pull request! Please provide enough information so that others can review your pull request. Link to relevant issues if possible.

For more information, see the [CONTRIBUTING guide](https://github.com/detekt/detekt/blob/main/.github/CONTRIBUTING.md).
-->

This PR adds a rule that checks for `@deprecated` tags in KDoc comments. These tags are [discouraged by the official Kotlin documentation as they work in Java but do nothing in Kotlin](https://kotlinlang.org/docs/kotlin-doc.html)

> KDoc does not support the `@deprecated` tag. Instead, please use the `@Deprecated` annotation.

This is a fairly important rule for people who come from Java and used `@deprecated` in their Javadoc. IntelliJ IDEA does not actually warn about this.

This is quite a simple rule too, feel free to tell me if I did anything wrong!

Also fixed a bug where `EndOfSentenceFormat` did not work on its own, and its tests relied on `KDocStyle` instead of `EndOfSentenceFormat`.

utybo commented:
I need to check if the generated website actually works on this, so this remains a draft atm.

I am also wondering if this rule could be integrated as part of the `KDocStyle` multirule in EndOfSentenceFormat.kt

BraisGabin commented on detekt-rules-documentation/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/documentation/DeprecatedBlockTag.kt:
```diff
@@ -0,0 +1,71 @@
+ * This rule reports use of the `@deprecated` block tag in KDoc comments. Deprecation must be specified using a
+ * `@Deprecated` annotation as adding a `@deprecated` block tag in KDoc comments has no effect. The `@Deprecated`
+ * annotation constructor has dedicated fields for a message and a type (warning, error, etc.). You can also use the
+ * `@ReplaceWith` annotation to specify how to solve the deprecation automatically via the IDE.
```
If there is official documentation talking about this a link to it would be great.

utybo replied:
The link would have to be a [text fragment](https://mgearon.com/html/text-fragments/) as there's no way to permalink to an infobox in the Kotlin documentation. Could be `https://kotlinlang.org/docs/kotlin-doc.html#8c82dc04:~:text=KDoc%20does%20not%20support%20the`

utybo replied:
Looks like that doesn't work either, it will have to be a link to the `@suppress` section right above it...

BraisGabin replied:
That's fine. Even if you link just the page is good.

BraisGabin commented on detekt-rules-documentation/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/documentation/DeprecatedBlockTag.kt:
```diff
@@ -0,0 +1,71 @@
+import io.gitlab.arturbosch.detekt.api.Severity
+import org.jetbrains.kotlin.psi.KtDeclaration
+
+// Note: this could probably be added to the KDocStyle class in EndOfSentenceFormat.kt
```
I agree. Can you do the change?

utybo replied:
I believe there is also a problem with `EndOfSentenceFormat`, as it does not have a `visitDeclaration` of its own and relies on `KDocStyle` calling it.

utybo replied:
Yep, that's also a bug, the tests for `EndOfSentenceFormat` are using `KDocStyle`, not `EndOfSentenceFormat`. I'll fix it in this PR.

BraisGabin replied:
You can remove this comment now, right?

utybo replied:
Yes, I just forgot to remeove it 🤦

utybo replied:
Done

BraisGabin reviewed (commented):
Nice Rule and nice PR :). Good test coverage 👍

utybo commented:
I fixed everything, I also noticed that that `EndOfSentenceFormat` relied on `KDocStyle` to call its verify and did not have a `visitDecleration` of its own, which I fixed as well.

I don't know if I should be the one resolving review conversations, so just in case I'm leaving them open :)

BraisGabin commented on detekt-rules-documentation/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/documentation/CommentSmellProvider.kt:
```diff
@@ -20,6 +20,7 @@ class CommentSmellProvider : DefaultRuleSetProvider {
             CommentOverPrivateFunction(config),
             CommentOverPrivateProperty(config),
             KDocStyle(config),
+            DeprecatedBlockTag(config),
```
If you add it in the KDocStyle MultiRule you don't need this line.

utybo replied:
Done

BraisGabin reviewed (approved):
👏👏👏

utybo commented:
I'm not entirely sure of what's going on with the code coverage, the report is showing changes that I haven't made 🤔

BraisGabin commented:
> I'm not entirely sure of what's going on with the code coverage, the report is showing changes that I haven't made 🤔

No porblem, that's a known issue of Codecov

utybo commented:
Are there any plans on merging this PR? I think it can be closed otherwise.

BraisGabin commented:
Sorry for the delay. Merged! Thanks again for the contribution :)
